### PR TITLE
Internal authorization via discovery service

### PR DESCRIPTION
**Pull request by byewokko**
- There is one private key saved in zookeeper, shared by all cluster ASAB services.
- All services use it to generate JWTokens for the authorization of internal requests.
- AuthService adds the internal private key into its known keys.

# Usage

To use the shiny new internal auth via discovery service, open a discovery session with argument `auth="internal"`:
```python
async with self.DiscoveryService.session(auth="internal") as session:
	async with session.get("http://asab-myservice-1.instance_id.asab") as response:
		...
```

To use the authorization from the incoming request (usually from the UI), open a discovery session with argument `auth=request`:
```python
async def some_handler(request):
	async with self.DiscoveryService.session(auth=request) as session:
		async with session.get("http://asab-myservice-1.instance_id.asab") as response:
			...
```

# TODO

- [x] AuthService must be able to run in "internal-only" mode without errors.
- [ ] Reload private key from zookeeper when it changes (`jwcrypto.jwt.JWTMissingKey` error)

### CODE REVIEW

File: asab/api/discovery.py [hunk 1]
```diff
@@ -20,6 +23,10 @@ def __init__(self, app, zkc, service_name="asab.DiscoveryService") -> None:
 		super().__init__(app, service_name)
 		self.ZooKeeperContainer = zkc
 		self.ProactorService = zkc.ProactorService
+		self.InternalAuthKeyPath = "/vault/internal_auth.pem"
```
Comment (ateska) [inline asab/api/discovery.py:26]: Not in vault, after all.

@eliska-n

File: asab/api/discovery.py [hunk 2]
```diff
@@ -31,11 +38,65 @@ def __init__(self, app, zkc, service_name="asab.DiscoveryService") -> None:
...
 
 	def _on_zk_ready(self, msg, zkc):
 		if zkc == self.ZooKeeperContainer:
 			self._update_cache(msg)
+			self.ProactorService.schedule(self._ensure_internal_auth_key, zkc)
+
+
+	def _ensure_internal_auth_key(self, zkc):
+		key = zkc.Client.get(self.InternalAuthKeyPath)
+		while not key:
```
Comment (ateska) [inline asab/api/discovery.py:52]: Please comment the principle here.

File: asab/api/discovery.py [hunk 3]
```diff
@@ -312,3 +402,19 @@ async def resolve(self, hostname: str, port: int = 0, family: int = socket.AF_IN
 
 class NotDiscoveredError(RuntimeError):
 	pass
+
+
+def _generate_private_key() -> bytes:
+	import cryptography.hazmat.backends
```
Comment (ateska) [inline asab/api/discovery.py:408]: This will introduce a dependency on cryptography module?

**Review by ateska — CHANGES_REQUESTED** [review]

File: asab/api/discovery.py (hunk 3 above)
Comment (byewokko) [reply asab/api/discovery.py:408]: i can make it optional, we already have it like that in [setup.py](https://github.com/TeskaLabs/asab/blob/master/setup.py#L83)
the dependency is already there in [asab.web.webcrypto](https://github.com/TeskaLabs/asab/blob/master/asab/web/webcrypto.py#L4) and **optionally** in [asab.storage](https://github.com/TeskaLabs/asab/blob/master/asab/storage/service.py#L9)

Comment (ateska) [reply asab/api/discovery.py:412]: Does `jwtcrypto` provide a key generation functions?

File: asab/api/discovery.py [hunk 4]
```diff
@@ -20,6 +23,10 @@ def __init__(self, app, zkc, service_name="asab.DiscoveryService") -> None:
 		super().__init__(app, service_name)
 		self.ZooKeeperContainer = zkc
 		self.ProactorService = zkc.ProactorService
+		self.InternalAuthKeyPath = "/asab/auth/internal_auth_private_key"
```
Comment (ateska) [inline asab/api/discovery.py:26]: `/asab/auth/internal_auth_private.key`

**Review by ateska — CHANGES_REQUESTED** [review]
Last remark

File: asab/api/discovery.py (hunk 3 above)
Comment (ateska) [reply asab/api/discovery.py:412]: discussed on review

File: asab/api/discovery.py [hunk 5]
```diff
@@ -31,11 +44,74 @@ def __init__(self, app, zkc, service_name="asab.DiscoveryService") -> None:
...
+		if self.InternalAuthToken:
+			if self.InternalAuthToken.claims.get("exp") > (
+				datetime.datetime.now(datetime.UTC) - datetime.timedelta(seconds=300)
+			):
+				# Token is valid and does not expire soon
+				return
+
+		claims = {
+			# Issuer (URL of the app that created the token)
+			"iss": "http://{instance_id}.{service_id}.asab",
```
Comment (byewokko) [inline asab/api/discovery.py:92]: @eliska-n  can i easily get the discovery URL of this very service?

**Comment by byewokko** [comment]
The state of this MR is that it is **NOT fully tested yet**. I only had time to test the initialization.

**Review by ateska — APPROVED** [review]

**Review by ateska — APPROVED** [review]

**Pull request merged**